### ISSUE TITLE
Using `pytest` for testing

### ISSUE THREAD

**Issue by pdobsan** [pdobsan/pynauty#7]
Although adding this feature is well under way I open this issue to avoid scattering  the related discussion all over in various threads. Also to summarize where we are now.

- Using the `pytest` framework is a good idea: kind of standard, thorough with good logging, easy to extend for our users etc.
- Virtaulenvs can cause problems, not just for `tests` but also for the `docs` target.
- Importing modules can be tricky, it seems necessary to check in advance a module availability and give the user a warning but not doing an automatic install behind their back.

The recent merge 59e1a4170a9b0255282239975738056f1ff164d5 solves (I hope) the virtualenv and module availability problems, see the details in the merge comments. This make unnecessary the make file trick in pull request #6. Also PR #6  has been made against `upstream/main` what I would rather not merge, so I close that. Please, create a pristine new branch out of my current `origin/main` for this purpose, say `tests-with-pytest` and create PR-s or drafts with this branch.

We leave it to the user to install and how to install a missing module, in this case `pytest`, but they might not want to or able to do this. So I suggest a creation of a very `minimal-test` which runs outside of `pytest`. Just checking importing `pynauty` and calling its functions with a simple graph(s). `tests` target would run first this `minimal-test` then attempt to run the full tests in `pytest`.

**Comment by goens** [comment]
I'm not sure if I understand the motivation for the minimal test. It seems redundant. I'm not sure why an user would not be able to install `pytest`, which is a very standard module. If they cannot install that, they probably also cannot install `pynauty`. So it seems to me more likely that they would not want to install `pytest`, not that they could not. In that case, I would expect them to also be able to choose to test it. 

Both cases sound like very fringe use-cases to me. I think of testing (and using a framework like `pytest`) for continuous integration. To make sure the package is always working during development and before uploading it. For using the package,  I would expect >90% of people to just run 
```
pip install pynauty
``` 
to get the package, and be done with this. That's the point of having a package manager like `pip` and a system like wheels. Then they should be sure that the binary they've downloaded works, which is why we should test it in CI. That's also my main motivation for having that in GitHub actions by the way, but that's a topic for the next PR/issue :)

Of course, it still makes sense to have a README and build system so that people can easily build and modify things on their own! I just wonder if it is worth the effort of thinking of fringe use-cases here, like someone who would want to build it on their own, but specifically choose not to install pynauty. I would think in that case, sure, you can do that, but then you're on your own. 

I'm not sure I understand the comment about the branches. Is the problem which branch I am sourcing this of in my repository (namely `main`), or which branch in your repository I'm suggesting to merge into? These two are completely independent. It doesn't matter what I call my branches, it doesn't change anything in your repository, so I'll reserve the right to work however I want to, especially if it does not change the outcome. If it's about the target branch, I'm confused as well. Why would you not want this change to come into the `main` branch? If you want to do it in steps, we can iterate on the PR, and you merge it only once you're satisfied with the changes. In the end, I do expect the changes to land in `main`, otherwise there's no point in doing them. By the way, this also applies to the "pristine" branch. It doesn't matter for the outcome how I do my commits. If you prefer you can even squash all of them into one single commit before merging, some people like that (I sincerely don't care), but for the outcome it doesn't matter how I choose to work for myself.

I'll make the PR once we've agreed on this. We clearly have our different workflows, and that's fine. But If I can be honest, it's frustrating if things just get closed and rejected instead of being discussed first, trying to solve the issues you have. I don't want that to happen again, so if I'd prefer discussing and agreeing first on what to do.

**Comment by pdobsan** [comment]
> I'm not sure if I understand the motivation for the minimal test. It seems redundant.

Okay, drop it then, it was just a suggestion.

**Comment by pdobsan** [comment]
> Both cases sound like very fringe use-cases to me. I think of testing (and using a framework like `pytest`) for continuous integration. To make sure the package is always working during development and before uploading it. For using the package, I would expect >90% of people to just run
> 
> ```
> pip install pynauty
> ```
> to get the package, and be done with this.

Yes, it's nice and we are getting there. But for many years people were using the package just fine without PyPi.

>That's the point of having a package manager like `pip` and a system like wheels. Then they should be sure that the binary they've downloaded works, which is why we should test it in CI. That's also my main motivation for having that in GitHub actions by the way, but that's a topic for the next PR/issue :)

I appreciate that, but even in GitHub actions my make based build method will be used as much as possible. I don't want to completely reprogram the dependency flow in actions' yaml or whatever. Having said that, I don't mind using this magic but it is not high on my priority list.

> Of course, it still makes sense to have a README and build system so that people can easily build and modify things on their own! I just wonder if it is worth the effort of thinking of fringe use-cases here, like someone who would want to build it on their own, but specifically choose not to install pynauty. I would think in that case, sure, you can do that, but then you're on your own.

I don't consider providing building `pynauty` from source caring for a fringe use-case as you repeatedly call it. Once again, this method worked for many years and I want to keep my independence from github CI magic as much as possible.

**Comment by goens** [comment]
Absolutely, I agree that building from source is an important use-case and great to make it as simple as possible! I'm sorry if my message was misleading: When I said it was a fringe use-case, I was talking specifically about a user that does not want to install `pytest` when building from source, not building from source in general! While I still think that, in general, many more people would just use binaries than build from source, if allowed the choice, that does not mean building from source is a fringe use-case, I would not say that at all.

About GitHub Actions, I guess that's also why we are doing this in two PRs, right? So, without GH Actions, what was missing in Pull Request #6 ? I'd like to make a new pull request with the pytest integration, but I don't understand the issue you have with the branches. Could you clarify that?

**Comment by pdobsan** [comment]
> While I still think that, in general, many more people would just use binaries than build from source, if allowed the choice, that does not mean building from source is a fringe use-case, I would not say that at all.

Yeah, just get a "binary" package installed from pypi sounds great. The reality is much much more complicated as I (re) learned while I managed to land the first beta on pypi. Python package management has changed a lot in recent years. Many solutions have grown out from the unpleasant situation with packaging, still the problems persist. Distributing mixed binary packages seemingly transparently serving many different platforms  is a mission impossible, only approximations can exist. So let be honest about it and give the building from sources method (in our case using make) the priority it deserves. The fact is, in general the users do not have a choice, the binary package just an illusion. Some times, maybe most of the times, it works, but ...

> About GitHub Actions, I guess that's also why we are doing this in two PRs, right?

Yes this is the main reason. There are other considerations I am not going into here.
Leave the actions alone for a while, please, until I ask for them.

Talking about packaging, publishing on pypi, there still are some frustrating open problems I am probably gonna open a separate issue about.

**Comment by pdobsan** [comment]
> So, without GH Actions, what was missing in Pull Request #6 ? I'd like to make a new pull request with the pytest integration, but I don't understand the issue you have with the branches. Could you clarify that?

Sigh, I can try, last time :-) There was nothing missing since the transformed pytest based tests were there and probably working. I got confused that the pr was made against upstream:main from fork:main instead of from a fork:topic branch. I asked around among git(hub) gurus, and they said it might be unusual and my confusion was understandable, but it is not a big deal. Then they went to show a dozen different and due to git terrible ux even more confusing ways how to deal with it.

I think a better way would be if a contributor's fork  is a well synchronized mirror of upstream's tree except the extra topic branch which is going to be the subject of the pr when it is done. The fork:main is just used to keep sucking in the changes from upstream. I try to keep the upstream:main in a production state all the time as much as possible. So that is the pilot line we keep ourself aligned to on both sides. The outlined method is a bit more balanced and shifts the burden of merge back to fork and make the cooperation more flexible and bidirectional.

For example, when the pr hits me and it works in the pr topic branch I can immediately accept it and merge into upstream:main. But I can also say, sorry although it works, but I am about to merge something (working on something) which affects the contents of the pr can we postpone and when I done adjust and test your topic branch and if there are things to work out please solve them on your side. Not to mention if the pr incomplete or does not work or not the way upstream would prefer. Or upstream may say, hey you work on topic-a but have a look at my topic-b somewhat related in progress can we exchange notes (or even patches)?

That is my mental model. As we know naming (and consequently renaming) is the most difficult problem in computing :-). So just humor me, will you?

**Comment by pdobsan** [comment]
> > I'm not sure if I understand the motivation for the minimal test. It seems redundant.
> 
> Okay, drop it then, it was just a suggestion.

May I backtrack on this a bit, mainly due to my experience in the last few days fighting with python's packaging. A minimal test is very handy while developing/experiencing with building workflow, just to check the building step (talking about compilation, linking etc.) it is not about the program correctness at that moment. The complete test suit takes too long as I was iterating it was annoying had to regularly kill it, then this kicks out the workflow from its track etc.
 
So I would still  like to have a mintest as described earlier but this should also be under the umbrella of pytest, I suppose that is easy to call only one dedicated test in certain cases. What do you think?

**Comment by goens** [comment]
> That is my mental model. As we know naming (and consequently renaming) is the most difficult problem in computing :-). So just humor me, will you?
Fair enough, I understand. Yep, sure :)
> Leave the actions alone for a while, please, until I ask for them.
I guess that's coming off the wrong way. That's why I was saying "it's for the next PR/issue", because I specifically did not want to continue discussing those in here.

I just opened the PR for this in #9
